File: 2022/zig/day2/src/day2p2.zig
const std = @import("std");
const utils = @import("utils");
const lib = @import("lib.zig");
const stdout = std.io.getStdOut().writer();

pub fn doit() !void {
    var buffer: [524288]u8 = undefined;
    var fba = std.heap.FixedBufferAllocator.init(&buffer);
    
    //var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
    var arena = std.heap.ArenaAllocator.init(fba.allocator());
    const alloc = arena.allocator();
    defer _ = arena.deinit();
    //errdefer _ = arena.deinit();

    var lines = try utils.get_lines("input", alloc);

    var rc: u64 = 0;
    for (lines.items) |line| {
        rc += get_score_line(line);
    }

    try stdout.print("Total: {d}\n", .{rc});
}

fn get_score_line(line: []const u8) u8 {
    const theirs = lib.RPSChoice.new(line[0]);
    const outcome = lib.DesiredOutcome.new(line[2]);
    const mine = outcome.get_choice(theirs);

    return lib.get_score(mine, theirs);
}

pub fn main() !void {
    var i: u32 = 0;
    while (i < 2500) : (i += 1) {
        try doit();
    }
}

File: 2022/rust/day2/src/day2p2.rs
use lib::*;
use utils::read_lines;

pub fn get_score_line(line: &str) -> u8 {
    let mut parser = line.chars();

    let theirs = RPSChoice::new(parser.next().unwrap());
    parser.next();
    let mine = DesiredOutcome::new(parser.next().unwrap());

    get_score(theirs, mine.get_rps_choice(theirs))
}

fn main() {
    let lines = read_lines("input").unwrap();

    let total = lines.fold(0, |acc, line| acc + get_score_line(&line.unwrap()) as u64);
    println!("total: {total}");
}
